### PR TITLE
Mention

### PR DESCRIPTION
**Pull request by sillych**
Support the new mention feature (Display Only)

2018-01-26-142618_832x268_scrot

will be

2018-01-26-142645_805x299_scrot

### CODE REVIEW

File: src/filters.ts [hunk 1]
```diff
@@ -126,6 +126,37 @@ angular.module('3ema.filters', [])
...
+            if (text !== null && text.length > 10) {
+                let result = text.match(/@\[([\*\@a-zA-Z0-9][\@a-zA-Z0-9]{7})\]/g);
+                if (result !== null) {
+                    result = ([...new Set(result)]);
+                    // Unique
+                    for (let possibleMention of result) {
+                        let identity = possibleMention.substr(2, 8);
+                        if (identity === '@@@@@@@@') {
+                            text = text.replace(
+                                new RegExp(possibleMention.replace(/([.*+?^=!:${}()|\[\]\/\\])/g, '\\$1'), 'g'),
```
Comment (dbrgn) [inline src/filters.ts]: Whoa, I don't understand this RegEx. What does it do?

Couldn't we simply do `text.replace('@[@@@@@@@@]', <span class="mention all">' + $translate.instant('messenger.ALL') + '</span>')` first, and only then search for contact mentions?

File: tests/filters.js [hunk 2]
```diff
@@ -121,4 +153,50 @@ describe('Filters', function() {
...
+                ['@[AAAAAAAA]', '<span class="mention contact">ContactA</span>'],
+                ['hello @[AAAAAAAA]. @[AAAAAAAA] you are my friend', 'hello <span class="mention contact">ContactA</span>. <span class="mention contact">ContactA</span> you are my friend'],
+                ['@[AAAAAAAA] @[AAAAAAAA] @[AAAAAAAA]', '<span class="mention contact">ContactA</span> <span class="mention contact">ContactA</span> <span class="mention contact">ContactA</span>']
+            ]);
+        });
+
+        it('mention - all', () => {
+            this.testPatterns([
+                ['@[@@@@@@@@]', '<span class="mention all">messenger.ALL</span>'],
+                ['@[@@@@@@@@] your base are belong to us', '<span class="mention all">messenger.ALL</span> your base are belong to us'],
```
Comment (dbrgn) [inline tests/filters.js]: I approve of this test case!

Comment (sillych) [reply tests/filters.js]: 👍

File: src/filters.ts (hunk 1 above)
Comment (sillych) [reply src/filters.ts]: this saves one iteration, replace will be only called if one or more mentions (including the `all` mention) are found

Comment (sillych) [reply src/filters.ts]: check my new commit

**Comment by sillych** [comment]
padding changed to `2px 5px;` 👾

**Comment by dbrgn** [comment]
When having mentions on two lines directly above each other, they partially overlap. But that's not too terrible, so it can still be fixed later :)

**Comment by rugk** [comment]
IMHO you should always still display the `@` char at the beginning. Also in the app, if it is not shown there.
Just so users know that this is a mention and see they can do the same by entering that character. Just how every other char app does it, so…

**Comment by ovalseven8** [comment]
> When having mentions on two lines directly above each other, they partially overlap. But that's not too terrible, so it can still be fixed later :)

Better create an issue :D

**Comment by mimiks** [comment]
> you should always still display the @ char at the beginning
The @ char looks pretty awkward. IMHO the grey frame is prominent enough.

**Comment by joelfischerr** [comment]
But the @ character is pretty much the universal character used for mentioning someone. I agree with @rugk on showing the '@' char at the beginning of every mention.

**Comment by mimiks** [comment]
It might indeed be helpful to new users. But on the other hand, displaying the @ character at the beginning of a mention is redundant and doesn't look good.

**Comment by rugk** [comment]
It is done like that in every chat application I know. So it confuses me a bit that this is not shown, even if I am not a beginner. 😄 
Also on GitHub it looks good this way, and seriously, no, there is no visual problem with an @ character. Or do you also don't like $ or # or what?

**Comment by mimiks** [comment]
> It is done like that in every chat application I know

Don't think so. Check out Telegram for instance. And even if it was, we don't have to do it same way.

**Comment by joelfischerr** [comment]
No, it's not redundant in my opinion as it clearly shows that someone is being mentioned. Just the highlighting of the text could lead to confusion as all names might be highlighted or it could be some formatting option. And as you can see in my reply above github shows the '@' character as well. I think it'd be better if it's done like everywhere else.

**Comment by lgrahl** [comment]
`@`, `#` & co don't work for me here because real names contain spaces and that looks awkward (and FWIW would be awkward to parse, too).

**Comment by rugk** [comment]
In MyBB it also allows spaces... But you already did the parsing work, so there is nothing to worry, it is just a visual change (i.e., it can even be accomplished with a CSS pseudo-class `::before`). (And I think the Threema IDs are used internally for the text, so it is not even a problem for the parser.)
It may look a bit surprising, but considering that you already have this grey background one clearly sees where the name starts and ends, i.e. it is just another character at the beginning, which is not a problem IMHO.

Maybe Threema can do an A/B test... 😄

**Comment by lgrahl** [comment]
Right, Threema IDs are used internally. The whole point of having an `@` in front of it (for me) is that I can copy and paste that text and it does the same thing - mention a person. Which it wouldn't in this case, so it makes no sense.

**Comment by joelfischerr** [comment]
It wouldn't preserve the mention when the text is copied back into Threema or when the text is copied elsewhere?
Because with the `@`in front of a name the meaning would be preserved in both cases. Without the `@`the meaning gets lost as soon as the text is shared.

> @, # & co don't work for me here because real names contain spaces and that looks awkward (and FWIW would be awkward to parse, too).

I don't think it looks awkward. The highlighting makes the block `@Max Musterman` look like one single entity.

**Comment by rugk** [comment]
> Because with the @in front of a name the meaning would be preserved in both cases. 

No, because internally it is `@ZU732H` or so, with the Threema ID. Threema just shows the name instead of the ID.
So either it needs to parse the name, too, in `@` statements (and here we get to the spaces problem), or you just cannot copy the element.

**Pull request merged**